extensions/resteasy-reactive/rest-client-microprofile/runtime/src/main/java/io/quarkus/resteasy/reactive/client/microprofile/MicroProfileRestRequestClientFilter.java

```diff
@@ -0,0 +1,96 @@
+package io.quarkus.resteasy.reactive.client.microprofile;
+
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.annotation.Priority;
+import javax.enterprise.context.RequestScoped;
+import javax.ws.rs.client.ClientRequestContext;
+import javax.ws.rs.client.ClientRequestFilter;
+import javax.ws.rs.core.MultivaluedHashMap;
+import javax.ws.rs.core.MultivaluedMap;
+
+import org.eclipse.microprofile.rest.client.ext.ClientHeadersFactory;
+import org.eclipse.microprofile.rest.client.ext.DefaultClientHeadersFactoryImpl;
+
+import io.quarkus.arc.Arc;
+import io.smallrye.common.constraint.NotNull;
+import io.smallrye.common.constraint.Nullable;
+
+@Priority(Integer.MIN_VALUE)
+public class MicroProfileRestRequestClientFilter implements ClientRequestFilter {
+    private static final MultivaluedMap<String, String> EMPTY_MAP = new MultivaluedHashMap<>();
+
+    @Nullable
+    private final HeaderFiller headerFiller;
+    @NotNull
+    private final ClientHeadersFactory headersFactory;
+
+    @Nullable
+    private final Method method;
+
+    /**
+     *
+     * @param headerFiller fills headers as specified in @ClientHeaderParam annotations
+     * @param headersFactory MP Rest Client headersFactory
+     * @param method java method of the JAX-RS interface
+     */
+    public MicroProfileRestRequestClientFilter(@Nullable HeaderFiller headerFiller,
+            @NotNull ClientHeadersFactory headersFactory,
+            // TODO: to optimize, add an option to disable passing the method?
+            @Nullable Method method) {
+        this.headerFiller = headerFiller;
+        this.headersFactory = headersFactory;
+        this.method = method;
+    }
+
+    // for each method, register one such filter
+    @Override
+    public void filter(ClientRequestContext requestContext) {
+        // mutable collection of headers
+        MultivaluedMap<String, String> headers = new MultivaluedHashMap<>();
+
+        // gather original headers
+        requestContext.getHeaders().forEach(
+                (key, values) -> headers.put(key, castToListOfStrings(values)));
+
+        // add headers from MP annotations
+        if (headerFiller != null) {
+            // add headers to a mutable headers collection
+            headerFiller.addHeaders(headers);
+        }
+
+        MultivaluedMap<String, String> incomingHeaders = MicroProfileRestRequestClientFilter.EMPTY_MAP;
+        if (Arc.container().getActiveContext(RequestScoped.class) != null) {
+            HeaderContainer headerContainer = Arc.container().instance(HeaderContainer.class).get();
```

Comment:
If this can't be injected it should be cached in a volatile field, so you don't have to do the lookup on every request.

Reply:
At the moment the filter instance is created per call. I plan to create a new PR with performance improvements, such as caching web targets, and reusing filters.
I will cache the header container proxy in it.